### DISCUSSION TITLE
Notifications to database.

### DISCUSSION THREAD

**Discussion by joe-hagerty** [spatie/laravel-backup#1525] [Q&A]
I log my notifications to a database, so I can then send a single notification digest at the end of the day.

Trying to configure database notifications I first tried just exchanging `mail` to `database` in the config.  But the default notifications do not have a toArray or toDatabase method, so I created my own notification, with a toArray function.  Now I am recieving an error message: 

```
In HasDatabaseNotifications.php line 14:

  Call to undefined method Spatie\Backup\Notifications\Notifiable::morphMany()
```

I believe this is because the database notification is expecting a user to be passed to it, and that I would need to further extend `Spatie\Backup\Notifications\Notifiable` but besides adding a morphMany() function, I am not sure what it is looking for.

Does anyone have any pointers?

Thanks

**Comment by joe-hagerty** [comment]
I have something that works, but I am not sure exactly why.  It turns out that the morphMany was not expecting a user, but a class which tells it which table to put the notifications in.

I had to add  to my custom BackupHasFailedNotification.php file.
```
    public function toArray()
    {
        return json_encode([
            "type" => "laravel-backup",
            "status"=> "failed",
            "content"=> trans('backup::notifications.backup_failed_subject')
        ]);
    }
```

Then to my CustomNotifiable.php class, I had to add
```
use Illuminate\Notifications\DatabaseNotification;
```
and
```
    public function routeNotificationForDatabase(): object
    {
        $notification = new DatabaseNotification;
        return $notification;
    }
```
This should probably be DatabaseNotification:class, but I am getting data into the database now.



just for clarity my config/backup.php now contains:
```
        'notifications' => [
		//\Spatie\Backup\Notifications\Notifications\BackupHasFailedNotification::class => [],
	        \App\Notifications\BackupHasFailedNotification::class => ['database'],
            \Spatie\Backup\Notifications\Notifications\UnhealthyBackupWasFoundNotification::class => [],
            \Spatie\Backup\Notifications\Notifications\CleanupHasFailedNotification::class => [],
            \Spatie\Backup\Notifications\Notifications\BackupWasSuccessfulNotification::class => [],
            \Spatie\Backup\Notifications\Notifications\HealthyBackupWasFoundNotification::class => [],
            \Spatie\Backup\Notifications\Notifications\CleanupWasSuccessfulNotification::class => [],
        ],

        /*
         * Here you can specify the notifiable to which the notifications should be sent. The default
         * notifiable will use the variables specified in this config file.
         */
        //'notifiable' => \Spatie\Backup\Notifications\Notifiable::class,
        'notifiable' => \App\Notifications\CustomNotifiable::class,
```

Obviously I still need to build the other Notifications.  Any other suggestions to help would be appreciated.

**Comment by ilyas-bensmaine** [comment]
I have been working on the same thing, i needed to send backup notifications via Database channel. 

Since the **spatie/laravel-backup** don't support database notification by default, i run into the same problems as you. So after few hours of research i found i gem article about ["how notification channels work" ](https://divinglaravel.com/how-notification-channels-work-in-laravel) written by this guy @[ Mohamed Said ](https://github.com/themsaid) which i highly recommend you to read it.

So my solution was too simple, by following the official documentation of **spatie/laravel-backup** package [(Adding extra notification channels)](https://spatie.be/docs/laravel-backup/v8/sending-notifications/adding-extra-notification-channels) and [(Customizing the notifiable)](https://spatie.be/docs/laravel-backup/v8/sending-notifications/customizing-the-notifiable) sections... The **routeNotificationForDatabase** method in your **CustomNotifiable.php** class should be something like this :
```php
use App\Models\User;
use Illuminate\Database\Eloquent\Relations\MorphMany;
use Spatie\Backup\Notifications\Notifiable;

class BackupNotifiable extends Notifiable
{
    public function routeNotificationForDatabase(): MorphMany
    {
        return User::find(1)->notifications();
    }
}
```
In my case i want to notify the user with ID 1, for you it can be any Notifiable instance you want.

The reason why this solution works because the `routeNotificationFor()` method which exists in the `Notifications\RoutesNotifications` trait, and used inside the `Notifications\Notifiable` trait.

However the `routeNotificationFor()` method is used to determine where to route the notification to, and it can  be override for **Database Driver**, by defining `routeNotificationForDatabase()` and it should return `: MorphMany` relationship as excepted in the original method in the code below :
```php
    /**
     * Get the notification routing information for the given driver.
     *
     * @param  string  $driver
     * @param  \Illuminate\Notifications\Notification|null  $notification
     * @return mixed
     */
    public function routeNotificationFor($driver, $notification = null)
    {
        if (method_exists($this, $method = 'routeNotificationFor'.Str::studly($driver))) {
            return $this->{$method}($notification);
        }

        return match ($driver) {
            'database' => $this->notifications(),
            'mail' => $this->email,
            default => null,
        };
    }
```

**Comment by NafiurSiddiqui** [comment]
Thanks guys, @ilyas-bensmaine @joe-hagerty I was stuck with this. Wish I had any time to work and make a pull request for this. I feel like 'database' channel should be included by deafult.
